### PR TITLE
Call Hierarchy

### LINKED ISSUE #2546: Implement Call Hierarchy
Let's implement https://github.com/microsoft/vscode-languageserver-node/blob/dbaeumer/moreIntegrationTests/protocol/src/protocol.callHierarchy.proposed.ts#L1

### PR DESCRIPTION
**Pull request by kjeremy**
Support experiment incoming and outgoing calls.

Fixes #2546

### CODE REVIEW

**Comment by kjeremy** [comment]
Blocked on #2708

**Comment by kjeremy** [comment]
@matklad As I'm writing tests for this I'm noticing that there is some non-determinism in between test runs. I think it's because I need to sort my `CallInfo` structs by `NavigationTarget` however this seems weird to me... do we do this anywhere else?

**Comment by memoryruins** [comment]
This is really neat!  Took a screenshot using your branch out of curiosity

screenshot 6249



> I'm not 100% convinced that I understand how this extension is supposed to work and have not compared it against any other implementations :)

and I am struggling to find other implementations! Only found a screenshot (UI concept?) and sample extension in https://code.visualstudio.com/updates/v1_33#_call-hierarchy

**Comment by kjeremy** [comment]
@memoryruins Awesome! I just pushed a change that allows me to find incoming calls all the way to `main`.

**Comment by kjeremy** [comment]
Here's the example they used to test this in vscode: https://github.com/jrieken/demo-callhierarchy

**Comment by kjeremy** [comment]
@matklad 

This is ready for review with inbound calls working. I will move protocol support over to lsp-types once I get outbound calls working but will do both of those in separate PRs.

**Comment by kjeremy** [comment]
Test failures seem unrelated (path conversions on windows?).

**Comment by kjeremy** [comment]
Outgoing calls are now implemented with tests.

**Comment by kjeremy** [comment]
I just updated everything to the newest lsp-types where I moved the protocol bits and squashed everything into one commit.

File: crates/ra_ide/src/call_hierarchy.rs [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+    goto_definition, references, FilePosition, NavigationTarget, RangeInfo,
+};
+
+#[derive(Default)]
+struct CallLocations {
+    funcs: IndexMap<NavigationTarget, Vec<TextRange>>,
+}
+
+impl CallLocations {
+    pub fn add(&mut self, target: &NavigationTarget, range: TextRange) {
```
Comment (matklad) [inline crates/ra_ide/src/call_hierarchy.rs]: These don't need to be `pub`. Also let's move this below `pub(crate)` function

Comment (kjeremy) [reply crates/ra_ide/src/call_hierarchy.rs]: > Also let's move this below `pub(crate)` function

You mean move them below `CallItem`'s impl block?

Comment (matklad) [reply crates/ra_ide/src/call_hierarchy.rs]: right, so that the public API of the module is on top

**Comment by mimoo** [comment]
hey! This doesn't work in vscode for me, am I missing something?

**Comment by kjeremy** [comment]
Define "doesn't work". Is it not showing in the context menu or is the UI
not showing calls into or from?

**Comment by mimoo** [comment]
oops, nevermind it works on some functions! (for context I have tried to use it several times in [libra](https://www.github.com/libra/libra) and most of the time it doesn't seem to work

**Comment by kjeremy** [comment]
It's powered by the find references functionality so I would be interested
to know if the functions where it fails succeeds there. A repro case would
be nice!

**Comment by mimoo** [comment]
ah I see, well this functionality seldom works in our codebase :)

**Comment by kjeremy** [comment]
@mimoo can you file an issue with an example of something you expect to work but doesn't? That would be really helpful!

**Comment by mimoo** [comment]
I'll look into it!

**Pull request merged**